examples/Atlas/+atlasControllers/@InstantaneousQPController/setupAndSolveQP.m

```diff
@@ -0,0 +1,355 @@
+% its own function. It also mirrors the functionality of instantaneousQPControllermex.cpp
+% @param t current time (s)
+% @param q, qd current robot pose and velocity
+% @param qp_input QPInput2D describing the current plan state
+% @param contact_sensor num_bodies vector indicating whether the given body has detected contact force
+
+params = obj.param_sets.(qp_input.param_set_name);
+r = obj.robot;
+ctrl_data = obj.controller_data;
+
+% Compute desired body accelerations and costs
+all_bodies_vdot = struct('body_id', cell(1,length(qp_input.body_motion_data)),...
+                         'body_vdot', cell(1,length(qp_input.body_motion_data)),...
+                         'params', cell(1,length(qp_input.body_motion_data)));
+fun = fcompare(@atlasControllers.bodyMotionControl,@instantaneousBodyMotionControlmex);
+for j = 1:length(qp_input.body_motion_data)
+  body_id = qp_input.body_motion_data(j).body_id;
+  [body_des, body_v_des, body_vdot_des] = evalCubicSplineSegment(t - qp_input.body_motion_data(j).ts(1), qp_input.body_motion_data(j).coefs);
+  body_vdot = fun(r, [q; qd], body_id,...
+                                body_des, body_v_des, body_vdot_des, params.body_motion(body_id));
+  all_bodies_vdot(j).body_id = body_id; 
+  all_bodies_vdot(j).body_vdot = body_vdot;
+  all_bodies_vdot(j).params = params.body_motion(body_id);
+end
+
+% Find the active set of supports
+mask = getActiveSupportsmex(obj.mex_ptr, [q; qd], qp_input.support_data, contact_sensor, params.contact_threshold, obj.default_terrain_height);
+supp = qp_input.support_data(logical(mask));
+
+% Run PD on the desired configuration
+qddot_des = obj.wholeBodyPID(t, q, qd, qp_input.whole_body_data.q_des, params.whole_body);
+
+% Find the constrained joints
+condof = qp_input.whole_body_data.constrained_dofs;
+
+% Extract the coefficient of friction
+if isempty(qp_input.support_data)
+  mu = 1;
+else
+  mu = qp_input.support_data(1).mu;
+  for j = 2:length(qp_input.support_data)
+    if qp_input.support_data(j).mu ~= mu
+      warning('Drake:MultipleMuNotSupported', 'We do not currently support different mu values for each support');
+    end
+  end
+end
+if (mu ~= 1)
+  r.warning_manager.warnOnce('Drake:MuAssumedToBe1', 'mu is assumed to be one in constactConstraints.m');
+end
+
+% Construct the linear system
+zmp_data = qp_input.zmp_data;
+B_ls = zmp_data.B;
+C_ls = zmp_data.C;
+D_ls = zmp_data.D;
+Qy = zmp_data.Qy;
+R_ls = zmp_data.R;
+S = zmp_data.S;
+s1 = zmp_data.s1;
+x0 = zmp_data.x0;
+y0 = zmp_data.y0;
+u0 = zmp_data.u0;
+w_qdd = params.whole_body.w_qdd;
+
+
+nq = r.getNumPositions();                                           
+kinsol = doKinematics(r,q,false,true,qd);
+
+R_DQyD_ls = R_ls + D_ls'*Qy*D_ls;
+
+for j = 1:length(supp)
+  supp(j).num_contact_pts = size(supp(j).contact_pts, 2);
+end
+
+dim = 3; % 3D
+nd = 4; % for friction cone approx, hard coded for now
+float_idx = 1:6; % indices for floating base dofs
```

Comment:
possibly dangerous with the upcoming quaternion floating base?  can we perhaps check this assumption in a constructor?